### PR TITLE
carddav: implement carddav.AddressBookQuery filtering

### PR DESCRIPTION
**Pull request by sbinet**
Fixes emersion/hydroxide#159.

Signed-off-by: Sebastien Binet <binet@cern.ch>

### CODE REVIEW

**Comment by sbinet** [comment]
just a first stab.

there are things I am not sure how to handle (the `IsNotDefined`s for example).
also this would need test(s).

File: carddav/carddav.go [hunk 1]
```diff
@@ -263,7 +263,143 @@ func (b *backend) ListAddressObjects(req *carddav.AddressDataRequest) ([]carddav
...
+			}
+		}
+		return true, nil
+	}
+}
+
+func matchPropFilter(prop carddav.PropFilter, ao carddav.AddressObject) (bool, error) {
+	field := ao.Card.Get(prop.Name)
+	if field == nil {
+		// assume AddressBookQuery.DataRequest.AllProp is false
```
Comment (emersion) [inline carddav/carddav.go:342]: `IsNotDefined` probably means that we need to check `field == nil`. Would need to read again the RFC to double check.

**Comment by emersion** [comment]
Looks like a god start! Did you have the chance to test it with a client that uses the request?

**Comment by sbinet** [comment]
yes, something like that:
```go
package main

import (
	"log"
	"net/http"

	"github.com/emersion/go-vcard"
	"github.com/emersion/go-webdav"
	"github.com/emersion/go-webdav/carddav"
)

func main() {
	cli, err := carddav.NewClient(
		webdav.HTTPClientWithBasicAuth(
			http.DefaultClient, "user", "s3cr3t",
		),
		"http://localhost:8080/",
	)
	if err != nil {
		log.Fatalf("could not create webdav client: %+v", err)
	}

	principal, err := cli.FindCurrentUserPrincipal()
	if err != nil {
		log.Fatalf("could not find principal: %+v", err)
	}
	log.Printf("principal: %q", principal)

	homeSet, err := cli.FindAddressBookHomeSet(principal)
	if err != nil {
		log.Fatalf("failed to query CardDAV address book home set: %v", err)
	}
	log.Printf("home-set: %q", homeSet)

	addressBooks, err := cli.FindAddressBooks(homeSet)
	if err != nil {
		log.Printf("failed to query CardDAV address books: %+v", err)
	}
	log.Printf("addr-books: %d", len(addressBooks))
	for i, book := range addressBooks {
		log.Printf("book[%d]: %+v", i, book)
	}

	addressBook := addressBooks[0]

	query := carddav.AddressBookQuery{
		DataRequest: carddav.AddressDataRequest{
			Props: []string{
				vcard.FieldFormattedName,
				vcard.FieldEmail,
				vcard.FieldUID,
			},
		},
		PropFilters: []carddav.PropFilter{{
			Name: vcard.FieldFormattedName,
		}},
	}

	queryText := "Binet"
	if queryText != "" {
		query.PropFilters = []carddav.PropFilter{
			{
				Name:        vcard.FieldFormattedName,
				TextMatches: []carddav.TextMatch{{Text: queryText}},
			},
			{
				Name:        vcard.FieldEmail,
				TextMatches: []carddav.TextMatch{{Text: queryText}},
			},
		}
	}

	addrs, err := cli.QueryAddressBook(addressBook.Path, &query)
	if err != nil {
		log.Printf("addrs: %d", len(addrs))
		log.Fatalf("failed to query CardDAV addresses: %+v", err)
	}
	log.Printf("addrs: %d", len(addrs))

	// TODO: cache the results
	emails := make([]string, 0, len(addrs))
	for _, addr := range addrs {
		cardEmails := addr.Card.Values(vcard.FieldEmail)
		emails = append(emails, cardEmails...)
		log.Printf("email: %q", cardEmails)
	}
}
```

I'll move the `matchXXX` stuff over to `vcard` and add tests there.
(then update this PR to use that new `Match` method)

**Comment by emersion** [comment]
Superseded by https://github.com/emersion/hydroxide/commit/1b5a2e71ba06b5e5c48107a8686e9d67703f32e8

**Pull request closed without merging**